File: js/programmers/level1/모의고사.js
function score(answers, p) {
  return answers.reduce(
    (acc, cur, idx) => acc + (p[idx % p.length] === cur && 1),
    0,
  );
}

function solution(answers) {
  const answer = [];

  const p1 = [1, 2, 3, 4, 5];
  const p2 = [2, 1, 2, 3, 2, 4, 2, 5];
  const p3 = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5];

  const s1 = score(answers, p1);
  const s2 = score(answers, p2);
  const s3 = score(answers, p3);

  const max = Math.max(s1, s2, s3);
  if (s1 === max) answer.push(1);
  if (s2 === max) answer.push(2);
  if (s3 === max) answer.push(3);

  return answer;
}

console.log(solution([1, 2, 3, 4, 5]));
console.log(solution([1, 3, 2, 4, 2]));

File: 프로그래머스/1/42840. 모의고사/모의고사.java
import java.util.*;

class Solution {
    public List<Integer> solution(int[] answers) {
        int[] scores = new int[3];
        
        int[] pattern1 = {1, 2, 3, 4, 5};
        int[] pattern2 = {2, 1, 2, 3, 2, 4, 2, 5};
        int[] pattern3 = {3, 3, 1, 1, 2, 2, 4, 4, 5, 5};
        
        for (int i = 0; i < answers.length; i++) {
            if (answers[i] == pattern1[i % pattern1.length]) {
                scores[0]++;
            }
            if (answers[i] == pattern2[i % pattern2.length]) {
                scores[1]++;
            }
            if (answers[i] == pattern3[i % pattern3.length]) {
                scores[2]++;
            }
        }
        
        int maxScore = Math.max(scores[0], Math.max(scores[1], scores[2]));
        
        List<Integer> result = new ArrayList<>();
        if (scores[0] == maxScore) result.add(1);
        if (scores[1] == maxScore) result.add(2);
        if (scores[2] == maxScore) result.add(3);
        
        return result;
    }
}
